src/Chapter3.hs

```diff
@@ -480,6 +509,50 @@ After defining the city, implement the following functions:
    and at least 10 living __people__ inside in all houses of the city in total.
 -}
 
+data Castle
+  = None
+  | Castle {castleName :: String}
+  | CastleWithWalls {castleName :: String}
+  deriving (Show)
+
+-- Why can't we use 1 | 2 | 3 | 4 here? 
+data HousePeople
+  = One
+  | Two
+  | Three
+  | Four
+  deriving (Show)
+
+data House = House {housePeople :: HousePeople} deriving (Show)
+
+data SpecialBuilding = Church | Library deriving (Show)
+
+data City = City {
+  cityCastle :: Castle,
+  citySpecialBuilding :: SpecialBuilding,
+  cityHouses :: [House]
+} deriving (Show)
+
+buildCastle :: City -> String -> City
+buildCastle city castleName = city {cityCastle = Castle castleName}
+
+buildHouse :: City -> House -> City
+buildHouse city house = city {cityHouses = house : cityHouses city}
+
+-- if we would have a nicer HousePeople constructor we could avoid this mess
+countPeople :: House -> Integer
+countPeople House {housePeople=One} = 1
+countPeople House {housePeople=Two} = 2
+countPeople House {housePeople=Three} = 3
+countPeople House {housePeople=Four} = 4
```

Comment:
You can write this function cleaner if you use the `case-of` expression:

```haskell
countPeople (House house) = case house of
    One -> 1
    Two -> 2
    ...
```

This is totally valid implementation 👍🏻 Moreover, compiler will warn you if you decide to extend the `HousePeople` data type in the future and forget to update this function.

If you don't want to do pattern matching at all, you can derive the `Enum` typeclass for `House` people and implement this function as a one-liner using `fromEnum`